### PR TITLE
add cond dce rule and custom-policy partial eval rule

### PR DESCRIPTION
**Pull request by mattjj**
This is the `cond` analogue of the new rule(s) for `scan` landed in #10576. Luckily `cond` is a lot easier!

Context from the PR message of #9148:
> A few months ago we wrote a new version of remat to support custom rematerialization policies. The new version is also built on a simpler basic implementation. But instead of replacing the pre-existing remat implementation, we kept both around because the new remat did not yet support higher-order primitives like scan and cond. Tech debt!

This PR adds two new rules for `cond`: the DCE rule and the `_cond_partial_eval_custom` rule. The latter is to support partial evaluation parameterized by a policy, so that what gets staged out is in general a superset of what's unknown: known things can get staged out too.

The tests here are basically analogues of the scan ones.

### CODE REVIEW

File: jax/_src/lax/control_flow/conditionals.py [hunk 1]
```diff
@@ -433,6 +435,78 @@ def _cond_partial_eval(trace, *tracers, branches, linear):
...
+  res_binders = map(newvar, all_res_avals)
+
+  # Build the known eqn.
+  ins_known, _ = partition_list(unks_in, eqn.invars)  # includes index invar
+  out_binders_known, _ = partition_list(unks_out, eqn.outvars)
+  linear_known = [l for l, uk in zip(eqn.params['linear'], ops_uk) if not uk]
+  params_known = dict(branches=branches_known, linear=tuple(linear_known))
+  eqn_known = pe.new_jaxpr_eqn(
+      ins_known, [*out_binders_known, *res_binders], cond_p, params_known,
+      core.no_effects, eqn.source_info)
```
Comment (sharadmv) [inline jax/_src/lax/control_flow/conditionals.py:497]: Should join the branch effects here and below, right?

File: jax/_src/lax/control_flow/conditionals.py [hunk 2]
```diff
@@ -509,15 +583,51 @@ def augment_jaxpr(jaxpr, res_indices):
...
+  # First, compute which inputs are used in any branch (not including `pred`).
+  used_inputs: List[bool] = [False] * (len(eqn.invars) - 1)  # -1 for pred
+  for jaxpr in branches:
+    _, used_inputs_ = pe.dce_jaxpr(jaxpr, used_outputs, instantiate=False)
+    used_inputs = map(operator.or_, used_inputs, used_inputs_)
+
+  # Next, compute DCEd branches, instantiating according to used_inputs.
+  dce_branches_ = [pe.dce_jaxpr(jaxpr, used_outputs, instantiate=used_inputs)[0]
+                   for jaxpr in branches]
+  dce_branches = [core.ClosedJaxpr(jaxpr, closed_jaxpr.consts)
```
Comment (sharadmv) [inline jax/_src/lax/control_flow/conditionals.py:608]: Are these consts always empty? If not, can they be DCEd?

File: jax/_src/lax/control_flow/conditionals.py [hunk 3]
```diff
@@ -509,15 +583,51 @@ def augment_jaxpr(jaxpr, res_indices):
...
+                  for closed_jaxpr, jaxpr in zip(closed_branches, dce_branches_)]
+
+  # Finally, update parameters and form the new eqn.
+  dce_linear = [l for l, used in zip(eqn.params['linear'], used_inputs) if used]
+  new_params = dict(eqn.params, branches=tuple(dce_branches),
+                    linear=tuple(dce_linear))
+  # TODO(mattjj,sharadmv): don't assume effects are never DCE'd?
```
Comment (sharadmv) [inline jax/_src/lax/control_flow/conditionals.py:615]: Can you not do a join across the branch effects here as well?

**Review by sharadmv — APPROVED** [review]

File: jax/_src/lax/control_flow/conditionals.py (hunk 1 above)
Comment (mattjj) [reply jax/_src/lax/control_flow/conditionals.py:497]: Oops, thanks! Forgot to update it to reflect #11641. Hoisted by my own petard.

File: jax/_src/lax/control_flow/conditionals.py (hunk 2 above)
Comment (mattjj) [reply jax/_src/lax/control_flow/conditionals.py:608]: Good q. It turns out they can't be DCE'd, in the sense that `dce_jaxpr` always leaves them untouched, just as a convention of that function. So this code is safe!

File: jax/_src/lax/control_flow/conditionals.py (hunk 3 above)
Comment (mattjj) [reply jax/_src/lax/control_flow/conditionals.py:615]: This was stale, not updated after https://github.com/google/jax/pull/11641. Will fix...

Comment (mattjj) [reply jax/_src/lax/control_flow/conditionals.py:615]: Ah actually, we can do a join here but `dce_jaxpr` will never eliminate effects. Really this TODO should go elsewhere, and I'll update the code here to be ready for `dce_jaxpr` potentially eliminating effects.

**Pull request merged**